### PR TITLE
util: support soft limit for memory tracker

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
Now the oomActions includes four types: 
1. rateLimitAction: limit the rate of some operators to reduce the memory usage, like TableReader 
2. spillAction: spilling intermediate data to reduce the memory usage.
3. panicAction: cancel the sql
4. logAction: log the sql, don't control the memory usage

Some of the Actions cannot immediately reduce memory usage, but alleviate the speed of memory growth. So I think we should separate logic to triage the Action. Introduce soft limit maybe a good way. (We can set soft limit as 80% of memory-quota by default)
1. When memory usage is higher than soft limit, we can triage the action that alleviate the speed of memory groth.
2. When memroy usage is higher than hard limit(memory-quota), we can triage the action that immediately reduce memory usage.

How it Works:

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- No release note.

### CODE REVIEW

File: util/memory/tracker.go [hunk 1]
```diff
@@ -255,18 +266,24 @@ func (t *Tracker) ReplaceChild(oldChild, newChild *Tracker) {
...
 	}
-	var rootExceed *Tracker
+	var rootExceed, rootExceedForSoftLimit *Tracker
 	for tracker := t; tracker != nil; tracker = tracker.getParent() {
-		if atomic.AddInt64(&tracker.bytesConsumed, bytes) >= tracker.bytesLimit && tracker.bytesLimit > 0 {
+		bytesConsumed := atomic.AddInt64(&tracker.bytesConsumed, bytes)
+		if bytesConsumed >= tracker.bytesLimit && tracker.bytesLimit > 0 {
 			rootExceed = tracker
 		}
+		if bytesConsumed >= int64(float64(tracker.bytesLimit)*softScale) && tracker.bytesLimit > 0 {
```
Comment (mmyj) [inline util/memory/tracker.go:284]: Maybe we could save the result of `float64(tracker.bytesLimit)*softScale` to a variable, and only update when `SetBytesLimit`?

Comment (wshwsh12) [reply util/memory/tracker.go]: Thanks for your suggestion. I have address this comment. PTAL @mmyj

**Review by mmyj — APPROVED** [review]
lgtm

File: util/memory/tracker.go [hunk 2]
```diff
@@ -46,22 +46,28 @@ type Tracker struct {
...
 	}
 
-	label         int   // Label of this "Tracker".
-	bytesConsumed int64 // Consumed bytes.
-	bytesLimit    int64 // bytesLimit <= 0 means no limit.
-	maxConsumed   int64 // max number of bytes consumed during execution.
-	isGlobal      bool  // isGlobal indicates whether this tracker is global tracker
+	label          int   // Label of this "Tracker".
+	bytesConsumed  int64 // Consumed bytes.
+	bytesLimit     int64 // bytesLimit <= 0 means no limit.
```
Comment (XuHuaiyu) [inline util/memory/tracker.go:58]: bytesHardLimit

File: util/memory/tracker.go [hunk 3]
```diff
@@ -39,97 +39,119 @@ import (
...
-	t.actionMu.actionOnExceed = &LogOnExceed{}
+	t.bytesSoftLimit = int64(float64(bytesLimit) * softScale)
+	t.actionMuForHardLimit.actionOnExceed = &LogOnExceed{}
 	t.isGlobal = true
 	return t
 }
 
 // CheckBytesLimit check whether the bytes limit of the tracker is equal to a value.
 // Only used in test.
 func (t *Tracker) CheckBytesLimit(val int64) bool {
```
Comment (mmyj) [inline util/memory/tracker.go:105]: Should rename the Checker/Setter/Getter?

File: util/memory/tracker.go [hunk 4]
```diff
@@ -317,8 +351,8 @@ func (t *Tracker) String() string {
 
 func (t *Tracker) toString(indent string, buffer *bytes.Buffer) {
 	fmt.Fprintf(buffer, "%s\"%d\"{\n", indent, t.label)
-	if t.bytesLimit > 0 {
-		fmt.Fprintf(buffer, "%s  \"quota\": %s\n", indent, t.FormatBytes(t.bytesLimit))
+	if t.bytesHardLimit > 0 {
```
Comment (mmyj) [inline util/memory/tracker.go:354]: Should we add a description for soft limit setting?

File: util/memory/tracker.go [hunk 5]
```diff
@@ -39,97 +39,119 @@ import (
...
 func (t *Tracker) FallbackOldAndSetNewAction(a ActionOnExceed) {
-	t.actionMu.Lock()
-	defer t.actionMu.Unlock()
-	t.actionMu.actionOnExceed = reArrangeFallback(t.actionMu.actionOnExceed, a)
+	t.actionMuForHardLimit.Lock()
+	defer t.actionMuForHardLimit.Unlock()
+	t.actionMuForHardLimit.actionOnExceed = reArrangeFallback(t.actionMuForHardLimit.actionOnExceed, a)
+}
+
+// FallbackOldAndSetNewActionForSoftLimit sets the action when memory usage exceeds soft bytesHardLimit
```
Comment (mmyj) [inline util/memory/tracker.go:142]: `soft bytesHardLimit`->`bytesSoftLimit`

File: util/memory/tracker.go [hunk 6]
```diff
@@ -39,113 +39,137 @@ import (
 // that is to say:
 // 1. Only "BytesConsumed()", "Consume()" and "AttachTo()" are thread-safe.
 // 2. Other operations of a Tracker tree is not thread-safe.
+//
+// We have two thresholds for the memory limit: soft limit and hard limit.
+// If the soft limit(0.8 * hard limit) is exceeded, we will trigger the action that alleviate the speed of memory growth.
```
Comment (XuHuaiyu) [inline util/memory/tracker.go:44]:
```suggestion
// If the soft threshold is exceeded, we will trigger the action that alleviates the speed of memory growth. The soft threshold is hard-coded as `0.8*hard threshold`. The actions that could be triggered are: [needs to be supplemented]
```

File: util/memory/tracker.go [hunk 7]
```diff
@@ -39,113 +39,137 @@ import (
...
 	}
 
-	label         int   // Label of this "Tracker".
-	bytesConsumed int64 // Consumed bytes.
-	bytesLimit    int64 // bytesLimit <= 0 means no limit.
-	maxConsumed   int64 // max number of bytes consumed during execution.
-	isGlobal      bool  // isGlobal indicates whether this tracker is global tracker
+	label          int   // Label of this "Tracker".
+	bytesConsumed  int64 // Consumed bytes.
+	bytesHardLimit int64 // bytesHardLimit <= 0 means no limit.
```
Comment (XuHuaiyu) [inline util/memory/tracker.go:62]: bytesHardThold
bytesSoftThold

File: util/memory/tracker.go [hunk 8]
```diff
@@ -39,113 +39,137 @@ import (
 // that is to say:
 // 1. Only "BytesConsumed()", "Consume()" and "AttachTo()" are thread-safe.
 // 2. Other operations of a Tracker tree is not thread-safe.
+//
+// We have two thresholds for the memory limit: soft limit and hard limit.
```
Comment (XuHuaiyu) [inline util/memory/tracker.go:43]:
```suggestion
// We have two thresholds for the memory quota: soft threshold and hard threshold.
```

File: util/memory/tracker.go [hunk 9]
```diff
@@ -39,113 +39,137 @@ import (
...
+// If the soft limit(0.8 * hard limit) is exceeded, we will trigger the action that alleviate the speed of memory growth.
+// If the hard limit(memory-quota-query) is exceeded, we will trigger the action that immediately reduce memory usage.
```
Comment (XuHuaiyu) [inline util/memory/tracker.go:45]:
```suggestion
// If the hard threshold is exceeded, we will trigger the action that immediately reduces memory usage. The hard threshold is set by the config item `mem-quota-query` or the system variable `tidb_mem_query_quota`. The actions that could be triggered are: [needs to be supplemented] 
```

File: util/memory/tracker.go [hunk 10]
```diff
@@ -39,113 +39,137 @@ import (
...
+	maxConsumed    int64 // max number of bytes consumed during execution.
+	isGlobal       bool  // isGlobal indicates whether this tracker is global tracker
+}
+
+type actionMu struct {
+	sync.Mutex
+	actionOnExceed ActionOnExceed
 }
 
+// softScale means the scale of soft limit to hard limit.
```
Comment (XuHuaiyu) [inline util/memory/tracker.go:73]:
```suggestion
// softScale means the scale of the soft threshold to the hard threshold.
```

File: util/memory/tracker.go [hunk 11]
```diff
@@ -39,113 +39,143 @@ import (
 // that is to say:
 // 1. Only "BytesConsumed()", "Consume()" and "AttachTo()" are thread-safe.
 // 2. Other operations of a Tracker tree is not thread-safe.
+//
+// We have two thresholds for the memory quota: soft threshold and hard threshold.
+// If the soft threshold is exceeded, we will trigger the action that alleviates the
+// speed of memory growth. The soft threshold is hard-coded as `0.8*hard threshold`.
+// The actions that could be triggered are: None.
+// If the hard threshold is exceeded, we will trigger the action that immediately
```
Comment (XuHuaiyu) [inline util/memory/tracker.go:47]:
```suggestion
//
// If the hard threshold is exceeded, we will trigger the action that immediately
```

**Review by XuHuaiyu — COMMENTED** [review]
LGTM

**Review by XuHuaiyu — APPROVED** [review]

File: util/memory/tracker.go (hunk 3 above)
Comment (wshwsh12) [reply util/memory/tracker.go:130]: I don’t think it’s necessary. I think hard/soft limit is just a concept inside tracker.

File: util/memory/tracker.go (hunk 4 above)
Comment (wshwsh12) [reply util/memory/tracker.go]: ditto.

**Pull request merged**